crates/bevy_macro_utils/src/static_ref.rs

```diff
@@ -0,0 +1,40 @@
+use proc_macro2::TokenStream;
+use quote::{quote, quote_spanned};
+use syn::spanned::Spanned;
+
+pub fn static_ref_impl(input: &syn::DeriveInput) -> TokenStream {
+    match &input.data {
+        syn::Data::Struct(data) => {
+            if data.fields.is_empty() {
+                quote! { ::std::option::Option::Some(&Self{}) }
+            } else {
+                quote! { ::std::option::Option::None }
+            }
+        }
+        syn::Data::Enum(data) => {
+            let variants: Vec<_> = data
+                .variants
+                .iter()
+                .filter_map(|variant| {
+                    let span = variant.span();
+                    let variant_ident = variant.ident.clone();
+                    if variant.fields.is_empty() {
+                        Some(quote_spanned! { span => Self::#variant_ident{} => ::std::option::Option::Some(&Self::#variant_ident{}), })
```

Comment:
Do we know for sure that

1. Those are unique per variant
2. Two different instances of `Foo::Bar.dyn_static_ref()` always return the same `&'static` reference? It's important since the whole point is to compare the reference instead of the type+value.

It seems like it, given we pass the tests, but we are relying on an implicit behavior.

Reply:
Regarding the second point, you are right, I was relying on implicit behavior. [The Rust Reference](https://doc.rust-lang.org/reference/items/constant-items.html#constant-items) states that 

> References to the same constant are not necessarily guaranteed to refer to the same memory address.

Therefore I changed the implementation to use [static items](https://doc.rust-lang.org/reference/items/static-items.html#static-items) instead, which solves this issue, because

> All references to the static refer to the same memory location.

Regarding the first point, I don't see a reason why two variants should share the same reference / memory address, since their representation in memory is different. Of course two different ZSTs can have the same memory address, but the cast to `&dyn` will yield different vtable pointers for different types, so they would not be considered equal.

Reply:
Also, because of the restrictions on [generics for static items](https://doc.rust-lang.org/reference/items/static-items.html#statics--generics), types with generic type or const parameters will not have the `static_ref` optimization and have to go through the `Interner`, but will still not cause an allocation if they are ZST.